### ISSUE TITLE
[Bug] Cannot read properties of undefined (reading 'shouldSupersede')

### ISSUE THREAD

**Issue by Turbo87** [emberjs/ember.js#20051]
### 🐞 Describe the Bug

When calling `routerService.urlFor(...)` the call throws the following error:

```
TypeError: Cannot read properties of undefined (reading 'shouldSupersede')
    at NamedTransitionIntent.applyToHandlers (http://localhost:4040/assets/vendor.js:58167:52)
    at NamedTransitionIntent.applyToState (http://localhost:4040/assets/vendor.js:58109:19)
    at PrivateRouter.generate (http://localhost:4040/assets/vendor.js:59134:26)
    at Router.generate (http://localhost:4040/assets/vendor.js:18718:38)
    at RouterService.urlFor (http://localhost:4040/assets/vendor.js:15630:27)
    at CatchAllRoute.urlForRouteInfo (webpack://cargo/./node_modules/ember-error-route/index.js?:84:24)
    at CatchAllRoute.serialize (webpack://cargo/./node_modules/ember-error-route/index.js?:60:21)
```

### 🔬 Minimal Reproduction

- Clone https://github.com/rust-lang/crates.io/pull/4659
- Run one of the failing tests

### 😕 Actual Behavior

The error page does not work properly.

### 🤔 Expected Behavior

The call does not fail and the error route is displayed correctly.

### 🌍 Environment

- Ember: 4.3.0 (works fine in 4.2.0)
- Node.js/npm: irrelevant
- OS: irrelevant
- Browser: irrelevant

### ➕ Additional Context

This is most likely related to the changes that were introduced in https://github.com/emberjs/ember.js/pull/19971

/cc @wagenet

**Comment by wagenet** [comment]
Ouch! Looks like we need more tests here. If you’re able to add some that would help speed up the debugging process.

**Comment by meirish** [comment]
We're seeing this in 3.28 fwiw.

**Comment by fsmanuel** [comment]
I have the same problem and could solve it by providing `@models` to `LinkTo`.

For example:
```hbs
<LinkTo @models={{array this.model}} @query={{hash state=@state page=1}}>Label</LinkTo>
```

**Comment by chriskrycho** [comment]
@meirish if you’re seeing it on 3.28 it’s *probably* a different specific bug. Can you create a new issue and a minimal reproduction? That would be super helpful.

@fsmanuel is that on 4.3+ or an earlier or later version where it first appears?

(Working on chipping away at this issue this week!)

**Comment by fsmanuel** [comment]
@chriskrycho if i recall it correctly it started with 3.27 and still persists in 3.28
Haven't updated to 4.x yet

**Comment by chriskrycho** [comment]
Thanks! As with @meirish's bug that's almost certainly a *different* bug than the one identified in this issue; if you can create a minimal reproduction that would be helpful. Thanks!

**Comment by fsmanuel** [comment]
@chriskrycho ah you are right. I'll see what I can do.

**Comment by chriskrycho** [comment]
I left a comment suggesting that we might have to revert #19971, but on digging further, I don't think that code path and this one intersect at all. It seems likelier that it's a result of the router.js refactors that landed in between. Investigating that now.

**Comment by chriskrycho** [comment]
It is indeed the result of the router.js work; tildeio/router.js#329 introduced a bug which was fixed by tildeio/router.js#334, which we actually have on the beta branch via #20166—I’ve tested that change (by editing `node_modules` in the `crates.io` repo :joy:) and it does indeed resolve it. I'll see if we can get the back-ports out tomorrow and this should be unblocked. 🎉

**Comment by Turbo87** [comment]
sweeeeeet! thank you, Chris! ❤️

**Comment by Turbo87** [comment]
I can confirm that CI is green now. thanks again!

**Comment by amk221** [comment]
I have just migrated from Ember's `LinkTo` to our own `Link` component, which hands off to the router service's `urlFor` method.

Since then, `TypeError: undefined is not an object (evaluating 'f.shouldSupersede')` has been showing up in our Sentry logs.

Could it possibly be related to https://github.com/emberjs/ember.js/issues/20663
